sys/kern/kern_zcond.c

```diff
@@ -0,0 +1,382 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <sys/types.h>
+#include <sys/param.h>
+#include <sys/systm.h>
+#include <sys/cpuset.h>
+#include <sys/eventhandler.h>
+#include <sys/kernel.h>
+#include <sys/linker.h>
+#include <sys/linker_set.h>
+#include <sys/lock.h>
+#include <sys/malloc.h>
+#include <sys/mutex.h>
+#include <sys/refcount.h>
+#include <sys/sbuf.h>
+#include <sys/smp.h>
+#include <sys/sysctl.h>
+#include <sys/zcond.h>
+
+#include <vm/vm.h>
+#include <vm/pmap.h>
+#include <vm/vm_extern.h>
+#include <vm/vm_kern.h>
+#include <vm/vm_map.h>
+#include <vm/vm_page.h>
+
+#include <machine/atomic.h>
+#include <machine/cpufunc.h>
+
+MALLOC_DECLARE(M_ZCOND);
+MALLOC_DEFINE(M_ZCOND, "zcond", "malloc for the zcond subsystem");
+
+static void
+zcond_load_patch_points(linker_file_t lf)
+{
+	struct patch_point *begin, *end;
+	struct patch_point *ins_p;
+	struct zcond *owning_zcond;
+
+	if (linker_file_lookup_set(lf, __XSTRING(ZCOND_LINKER_SET), &begin,
+		&end, NULL) == 0) {
+		for (ins_p = begin; ins_p < end; ins_p++) {
+			owning_zcond = ins_p->zcond;
+
+			if (owning_zcond->patch_points.slh_first == NULL) {
+				SLIST_INIT(&owning_zcond->patch_points);
+			}
+
+			SLIST_INSERT_HEAD(&owning_zcond->patch_points, ins_p,
+			    next);
+		}
+	}
+}
+
+static void
+zcond_kld_load(void *arg __unused, struct linker_file *lf)
+{
+	zcond_load_patch_points(lf);
+}
+
+static int
+zcond_load_patch_points_cb(linker_file_t lf, void *arg __unused)
+{
+	zcond_load_patch_points(lf);
+	return (0);
+}
+
+/*
+ * Collect patch_points from the __zcond_table ELF section into a list.
+ * Prepare a CPU local copy of the kernel_pmap, used to safely patch
+ * an instruction.
+ */
+static vm_offset_t mirror_addr;
```

Comment:
```suggestion
static vm_offset_t patch_vaddr; // ili nešto slično
```
Gurnuti na vrh, prije deklaracije static funkcija. Možda isto dodati komentar čemu služi.